kotlinx-coroutines-core/common/src/channels/BroadcastChannel.kt

```
@@ -70,6 +74,333 @@ public fun <E> BroadcastChannel(capacity: Int): BroadcastChannel<E> =
+        // as the operation should wait on multiple subscribers simultaneously.
+        // At the same time, broadcasts are obsolete, so we need a simple implementation
+        // that works somehow. Here is a tricky work-around. First, we launch a new
+        // coroutine that performs plain `send(..)` operation and tries to complete
+        // this `select` via `trySelect`, independently on whether it is in the
+        // registration or in the waiting phase. On success, the operation finishes.
+        // On failure, if another clause is already selected or the `select` operation
+        // has been cancelled, we observe non-linearizable behaviour, as this `onSend`
+        // clause is completed as well. However, we believe that such a non-linearizability
+        // is fine for obsolete API. The last case is when the `select` operation is still
+        // in the registration case, so this `onSend` clause should be re-registered.
+        // The idea is that we keep information that this `onSend` clause is already selected
+        // and finish immediately.
+        @Suppress("UNCHECKED_CAST")
+        element as E
+        // First, check whether this `onSend` clause is already
+        // selected, finishing immediately in this case.
+        lock.withLock {
+            val result = onSendInternalResult.remove(select)
+            if (result != null) { // already selected!
+                // `result` is either `Unit` ot `CHANNEL_CLOSED`.
+                select.selectInRegistrationPhase(result)
+                return
+            }
+        }
+        // Start a new coroutine that performs plain `send(..)`
+        // and tries to select this `onSend` clause at the end.
+        CoroutineScope(select.context).launch(start = CoroutineStart.UNDISPATCHED) {
+            val success: Boolean = try {
+                send(element)
+                // The element has been successfully sent!
+                true
+            } catch (t: Throwable) {
```

Please don't ever catch non-rethrowing `Throwable` in libraries code.
Consider catching `send`-specific exception type instead

---

Channels may be canceled with any exception using the deprecated API:
```
 @Deprecated(level = DeprecationLevel.HIDDEN, message = "Binary compatibility only")
    public fun cancel(cause: Throwable? = null): Boolean
```

To provide compatibility with such code, we need to catch all exceptions.

---

They can also be closed with an arbitrary exception concurrently, and also there is `broadcasst {}` builder.
Apart from that, there are system-wide throwables -- OOM, SoE, VerifyError etc. 
The point is, it would be nice to avoid `catch (Throwable)` and use any other more granular mechanism

---

What do you think about this fix? It makes the code much safer, neither breaking the compatibility nor requiring a special `send(..)` implementation.

```
            } catch (t: Throwable) {
                // This broadcast must be closed. However, it is possible that
                // an unrelated exception, such as `OutOfMemoryError` has been thrown.
                // This implementation checks that the channel is actually closed,
                // re-throwing the caught exception otherwise.
                if (isClosedForSend) false
                else throw t
            }
```

---

It probably should be `if (isClosedForSend && closeCause === unwrap(t))`